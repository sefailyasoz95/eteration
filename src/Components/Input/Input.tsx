import React, { useRef, useState } from "react";
import {
	View,
	TextInput,
	KeyboardType,
	StyleSheet,
	Animated,
	Easing,
	StyleProp,
	ViewStyle,
	TextInputProps,
	Platform,
} from "react-native";

type Props = {
	placeholder: string;
	keyboardType?: KeyboardType;
	autoFocus?: boolean;
	required?: boolean;
	hasError?: boolean;
	corner?: "cornered" | "curved" | "rounded" | "circle" | number;
	type?: "input" | "textarea";
	onTextChanged: Function;
	style?: StyleProp<ViewStyle> | any;
	testID: string;
};

const Input: React.FC<Props> = ({
	placeholder,
	keyboardType,
	autoFocus,
	hasError,
	required,
	corner,
	type,
	onTextChanged,
	testID,
}) => {
	const [inputError, setInputError] = useState(false);
	const placeholderRef = useRef(new Animated.Value(11)).current;
	const inputRef = useRef<TextInput | null>(null);
	const inputContainerStyles: Object = {
		borderRadius:
			corner === "cornered"
				? 0
				: corner === "curved"
				? 5
				: corner === "rounded"
				? 10
				: corner === "circle"
				? "50%"
				: corner,
		borderColor: (required && inputError) || hasError ? "red" : !inputError && "black",
		height: type === "textarea" ? 80 : 40,
	};
	const inputStyles: Object = {};
	const handleInputValue = (text: string | number) => {
		if (!text) {
			setInputError(true);
			Animated.timing(placeholderRef, {
				useNativeDriver: true,
				easing: Easing.bounce,
				duration: 200,
				toValue: 11,
			}).start();
		} else {
			setInputError(false);
			onTextChanged(text);
		}
	};
	return (
		<View
			style={[styles.inputContainer, inputContainerStyles]}
			onTouchEnd={() => inputRef.current?.focus()}
			testID='inputContainer'>
			<Animated.Text style={[styles.placeholder, { transform: [{ translateY: placeholderRef }] }]}>
				{placeholder}
			</Animated.Text>
			<TextInput
				ref={inputRef}
				testID={testID || "textinput"}
				style={[styles.input, inputStyles]}
				keyboardType={keyboardType}
				multiline={type === "textarea" && true}
				onEndEditing={(e) => {
					handleInputValue(e.nativeEvent.text);
				}}
				onFocus={() => {
					Animated.timing(placeholderRef, {
						useNativeDriver: true,
						easing: Easing.ease,
						duration: 200,
						toValue: -20,
					}).start();
				}}
				autoFocus={autoFocus}
			/>
		</View>
	);
};
const styles = StyleSheet.create({
	inputContainer: {
		borderStyle: "solid",
		borderWidth: 1,
		width: "95%",
		marginVertical: 12,
	},
	input: {
		paddingTop: Platform.OS === "ios" ? 8 : 2,
		paddingHorizontal: 8,
		marginTop: 3,
	},
	placeholder: {
		color: "rgba(0,0,0,0.5)",
		position: "absolute",
		marginLeft: 5,
		fontSize: 12,
	},
});
export default Input;
