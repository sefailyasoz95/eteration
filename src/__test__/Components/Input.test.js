import React from "react";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { render } from "@testing-library/react-native";
import Input from "../../Components/Input/Input";
import { create } from "react-test-renderer";

configure({ adapter: new Adapter() });
const event = {
	nativeEvent: {
		text: "testing input",
	},
};

describe("Input Component", () => {
	it("should render without issues", () => {
		const component = shallow(<Input />);
		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});
	it("checking onEndEditing functionality", () => {
		const onTextChanged = jest.fn();
		const setInputError = jest.fn();
		const component = create(<Input onTextChanged={onTextChanged} />);
		const checkState = jest.spyOn(React, "useState");
		const input = component.root.findByProps({ testID: "textinput" }).props;
		input.onEndEditing(event);
		checkState.mockImplementation((inputError) => [inputError, setInputError]);
		expect(setInputError).toBeTruthy();
		expect(onTextChanged).toBeCalledWith("testing input");
	});
});
