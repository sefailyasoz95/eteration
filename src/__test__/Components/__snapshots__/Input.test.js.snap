// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should render without issues 1`] = `
<View
  onTouchEnd={[Function]}
  style={
    Array [
      Object {
        "borderStyle": "solid",
        "borderWidth": 1,
        "marginVertical": 12,
        "width": "95%",
      },
      Object {
        "borderColor": "black",
        "borderRadius": undefined,
        "height": 40,
      },
    ]
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    style={
      Array [
        Object {
          "color": "rgba(0,0,0,0.5)",
          "fontSize": 12,
          "marginLeft": 5,
          "position": "absolute",
        },
        Object {
          "transform": Array [
            Object {
              "translateY": 11,
            },
          ],
        },
      ]
    }
  />
  <TextInput
    allowFontScaling={true}
    multiline={false}
    onEndEditing={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "marginTop": 3,
          "paddingHorizontal": 8,
          "paddingTop": 8,
        },
        Object {},
      ]
    }
    testID="textinput"
    underlineColorAndroid="transparent"
  />
</View>
`;
